ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterTableFunctionTransposeRule.java

```diff
@@ -0,0 +1,183 @@
+
+    for (RexNode ce : RelOptUtil.conjunctions(filterRel.getCondition())) {
+      Set<Integer> inputRefs = HiveCalciteUtil.getInputRefs(ce);
+
+      boolean canBePushed = true;
+      for (Integer inputRef : inputRefs) {
+        if (inputRef >= numFieldsInInput) {
+          canBePushed = false;
+          break;
+        }
+      }
+
+      if (canBePushed) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public void onMatch(RelOptRuleCall call) {
+    final Filter filter = call.rel(0);
+    final HiveTableFunctionScan tfs = call.rel(1);
+    final RelNode inputRel = tfs.getInput(0);
+    final int numFieldsInInput = inputRel.getRowType().getFieldCount();
+
+    final List<RexNode> newPartKeyFilterConditions = new ArrayList<>();
+    final List<RexNode> unpushedFilterConditions = new ArrayList<>();
+
+    // Check for each individual 'and' condition so that we can push partial
+    // expressions through.
+    for (RexNode ce : RelOptUtil.conjunctions(filter.getCondition())) {
+      Set<Integer> inputRefs = HiveCalciteUtil.getInputRefs(ce);
+      boolean canBePushed = true;
+      // We can only push if all the InputRef pointers are referencing the
+      // input RelNode to the TableFunctionScan
+      for (Integer inputRef : inputRefs) {
+        if (inputRef >= numFieldsInInput) {
+          canBePushed = false;
+          break;
+        }
+      }
+      if (canBePushed) {
+        newPartKeyFilterConditions.add(ce);
+      } else {
+        unpushedFilterConditions.add(ce);
+      }
+    }
+
+    // The "matches" check should guarantee there's something to push.
+    Preconditions.checkState(!newPartKeyFilterConditions.isEmpty());
+    final RexNode filterCondToPushBelowProj = RexUtil.composeConjunction(
+        filter.getCluster().getRexBuilder(), newPartKeyFilterConditions, true);
+
+    // Create the new filter with the pushed through conditions
+    final RelNode newFilter =
+        filter.copy(filter.getTraitSet(), tfs.getInput(0), filterCondToPushBelowProj);
+
+    // If there are conditions that cannot be pushed through, generate the RexNode
+    final RexNode unpushedFilCondAboveProj = unpushedFilterConditions.isEmpty()
+        ? null
+        : RexUtil.composeConjunction(filter.getCluster().getRexBuilder(),
+            unpushedFilterConditions, true);
+
+    // Generate the new TableFunctionScanNode with the Filter InputRel
+    final RelNode tableFunctionScanNode = tfs.copy(tfs.getTraitSet(), ImmutableList.of(newFilter),
+        tfs.getCall(), tfs.getElementType(), tfs.getRowType(), tfs.getColumnMappings());
+
+    // If there are expressions that couldn't be pushed through, generate the filter above the
+    // TableFunctionScan with these conditions.
+    final RelNode topLevelNode = unpushedFilCondAboveProj == null
+        ? tableFunctionScanNode
+        : filter.copy(filter.getTraitSet(), tableFunctionScanNode, unpushedFilCondAboveProj);
```

Comment:
Not sure if I applied this correctly.  I couldn't use the relBuilder to build the HiveLVTableFunctionScan, so I kept it the same way.

I suppose I could override the "functionScan()" method in HiveRelBuilder?  Didnt' feel great about doing that since we now have both HiveTableFunctionScan and HiveLVTableFunctionScan, but maybe I should?  Thoughts?